opentracing-spring-web-handler-interceptor/src/main/java/io/opentracing/contrib/spring/web/interceptor/TracingHandlerInterceptor.java

```diff
@@ -0,0 +1,147 @@
+            log.severe("Span not found! TracingFilter is not registered!?");
+            return true;
+        }
+
+        // skip preHandler after afterConcurrentHandlingStarted
+        if (httpServletRequest.getAttribute(AFTER_CONCURRENT_HANDLING_STARTED) != null) {
+            httpServletRequest.removeAttribute(AFTER_CONCURRENT_HANDLING_STARTED);
+            return true;
+        }
+
+        SpanWrapper filterSpan = (SpanWrapper) spanAttribute;
+
+        Span serverSpan = filterSpan.get();
+
+        /**
+         * If span in filter is finished create new span. However we need to check if there is already created span
+         * for this use case.
+         */
+        if (filterSpan.isFinished()) {
+            Object continuedSpanAttribute = httpServletRequest.getAttribute(CONTINUED_SERVER_SPAN);
+            if (continuedSpanAttribute == null ||
+                    (continuedSpanAttribute instanceof InterceptorSpanWrapper &&
+                            ((InterceptorSpanWrapper)continuedSpanAttribute).isFinished())) {
+                serverSpan = tracer.buildSpan(httpServletRequest.getMethod())
+                        .addReference(References.FOLLOWS_FROM, filterSpan.get().context())
+                        .start();
+                httpServletRequest.setAttribute(CONTINUED_SERVER_SPAN, new InterceptorSpanWrapper(serverSpan, handler));
+                // override server span context so local span in controllers will be child of it
+                httpServletRequest.setAttribute(TracingFilter.SERVER_SPAN_CONTEXT, serverSpan.context());
+            }
+        }
+
+        for (SpanDecorator decorator: decorators) {
+            decorator.onPreHandle(httpServletRequest, handler, serverSpan);
+        }
+
+        return true;
+    }
+
+    @Override
+    public void afterConcurrentHandlingStarted(HttpServletRequest request, HttpServletResponse response,
+                                               Object handler) throws Exception {
+        request.setAttribute(AFTER_CONCURRENT_HANDLING_STARTED, Boolean.TRUE);
+    }
+
+    @Override
+    public void afterCompletion(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse,
+                                Object handler, Exception ex) throws Exception {
+
+        Object spanAttribute = httpServletRequest.getAttribute(TracingFilter.SERVER_SPAN);
+        if (!(spanAttribute instanceof SpanWrapper)) {
+            log.severe("Span not found! TracingFilter is not registered!?");
+            return;
+        }
+
+        SpanWrapper filterSpan = (SpanWrapper) spanAttribute;
+
+        Span serverSpan = filterSpan.get();
+        if (filterSpan.isFinished()) {
+            serverSpan = ((InterceptorSpanWrapper)httpServletRequest.getAttribute(CONTINUED_SERVER_SPAN)).get();
+        }
+
+        for (SpanDecorator decorator: decorators) {
+            decorator.onAfterCompletion(httpServletRequest, httpServletResponse, handler, ex, serverSpan);
```

Comment:
Same here: what happens if a decorator throws an exception? Should it propagate to the end user?